File: VB/SummaryCustomDataBinding/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Data
Imports DevExpress.Web.Mvc
Imports SummaryCustomDataBinding.Models
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc
Imports System.Text.RegularExpressions
Namespace SummaryCustomDataBinding.Controllers
	Public Class HomeController
		Inherits Controller
		'
		' GET: /Home/

		Public Function Index() As ActionResult
			Return View()
		End Function
		Public Function AdvancedCustomBindingPartial(ByVal customCommand As String, ByVal fieldName As String) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("gridView")

			If viewModel Is Nothing Then
				viewModel = CreateGridViewModelWithSummary()
			End If
			If (Not String.IsNullOrWhiteSpace(customCommand)) AndAlso (Not String.IsNullOrWhiteSpace(fieldName)) Then
				ModifySummaries(viewModel, customCommand, fieldName)
			End If
			Return AdvancedCustomBindingCore(viewModel)
		End Function
		Public Sub ModifySummaries(ByVal viewModel As GridViewModel, ByVal command As String, ByVal fieldName As String)
			Dim shortCommand As String = Regex.Match(command, "Summary(.*)").Groups(1).Value

			If shortCommand = "None" Then
				viewModel.TotalSummary.RemoveAll(Function(item) item.FieldName = fieldName)
				Return
			End If
			Dim result As SummaryItemType
			If System.Enum.TryParse(Of SummaryItemType)(shortCommand, result) Then
                If (Not viewModel.TotalSummary.Exists(Function(item) item.FieldName = fieldName AndAlso item.SummaryType = result)) Then
                    viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = fieldName, .SummaryType = result})
                Else
                    viewModel.TotalSummary.RemoveAll(Function(item) item.FieldName = fieldName AndAlso item.SummaryType = result)
                End If

			End If
		End Sub
		' Paging
		Public Function AdvancedCustomBindingPagingAction(ByVal pager As GridViewPagerState) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("gridView")
			viewModel.ApplyPagingState(pager)
			Return AdvancedCustomBindingCore(viewModel)
		End Function
		' Filtering
		Public Function AdvancedCustomBindingFilteringAction(ByVal filteringState As GridViewFilteringState) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("gridView")
			viewModel.ApplyFilteringState(filteringState)
			Return AdvancedCustomBindingCore(viewModel)
		End Function
		' Sorting
		Public Function AdvancedCustomBindingSortingAction(ByVal column As GridViewColumnState, ByVal reset As Boolean) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("gridView")
			viewModel.ApplySortingState(column, reset)
			Return AdvancedCustomBindingCore(viewModel)
		End Function
		' Grouping
		Public Function AdvancedCustomBindingGroupingAction(ByVal column As GridViewColumnState) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("gridView")
			viewModel.ApplyGroupingState(column)
			Return AdvancedCustomBindingCore(viewModel)
		End Function

		Private Function AdvancedCustomBindingCore(ByVal viewModel As GridViewModel) As PartialViewResult
            viewModel.ProcessCustomBinding(AddressOf GridViewCustomBindingHandlers.GetDataRowCountAdvanced, AddressOf GridViewCustomBindingHandlers.GetDataAdvanced, AddressOf GridViewCustomBindingHandlers.GetSummaryValuesAdvanced, AddressOf GridViewCustomBindingHandlers.GetGroupingInfoAdvanced, AddressOf GridViewCustomBindingHandlers.GetUniqueHeaderFilterValuesAdvanced)
			Return PartialView("_GridViewPartial", viewModel)
		End Function

		Private Shared Function CreateGridViewModelWithSummary() As GridViewModel
			Dim viewModel = New GridViewModel()
			viewModel.KeyFieldName = "OrderID"

			viewModel.Columns.Add("OrderID")
			viewModel.Columns.Add("EmployeeID")
			viewModel.Columns.Add("OrderDate")
			viewModel.Columns.Add("Freight")
			viewModel.Columns.Add("ShipName")
			viewModel.Columns.Add("ShipAddress")
			viewModel.Columns.Add("ShipCity")
			viewModel.Columns.Add("ShipCountry")

			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "OrderID", .SummaryType = SummaryItemType.Sum})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "EmployeeID", .SummaryType = SummaryItemType.Count})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "Freight", .SummaryType = SummaryItemType.Average})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "ShipCity", .SummaryType = SummaryItemType.Min})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "OrderDate", .SummaryType = SummaryItemType.Min})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "Freight", .SummaryType = SummaryItemType.Count})

			Return viewModel
		End Function

	End Class

End Namespace

File: CS/SummaryCustomDataBinding/Controllers/HomeController.cs
using DevExpress.Data;
using DevExpress.Web.Mvc;
using SummaryCustomDataBinding.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using System.Text.RegularExpressions;
namespace SummaryCustomDataBinding.Controllers
{
    public class HomeController : Controller
    {
        //
        // GET: /Home/

        public ActionResult Index()
        {
            return View();
        }
        public ActionResult AdvancedCustomBindingPartial(string customCommand, string fieldName)
        {
            var viewModel = GridViewExtension.GetViewModel("gridView");

            if (viewModel == null)
                viewModel = CreateGridViewModelWithSummary();
            if (!String.IsNullOrWhiteSpace(customCommand) && !String.IsNullOrWhiteSpace(fieldName))
                ModifySummaries(viewModel, customCommand, fieldName);
            return AdvancedCustomBindingCore(viewModel);
        }
        public void ModifySummaries(GridViewModel viewModel, string command, string fieldName)
        {
            string shortCommand = Regex.Match(command, "Summary(.*)").Groups[1].Value;
            
            if (shortCommand == "None")
            {
                viewModel.TotalSummary.RemoveAll(item => item.FieldName == fieldName);
                return;
            }
            SummaryItemType result;
            if (Enum.TryParse<SummaryItemType>(shortCommand, out result))
            {
                if (!viewModel.TotalSummary.Exists(item => item.FieldName == fieldName && item.SummaryType == result))
                {
                    viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = fieldName, SummaryType = result });
                }
                else
                {
                    viewModel.TotalSummary.RemoveAll(item => item.FieldName == fieldName && item.SummaryType == result);
                }

            }
        }
        // Paging
        public ActionResult AdvancedCustomBindingPagingAction(GridViewPagerState pager)
        {
            var viewModel = GridViewExtension.GetViewModel("gridView");
            viewModel.ApplyPagingState(pager);
            return AdvancedCustomBindingCore(viewModel);
        }
        // Filtering
        public ActionResult AdvancedCustomBindingFilteringAction(GridViewFilteringState filteringState)
        {
            var viewModel = GridViewExtension.GetViewModel("gridView");
            viewModel.ApplyFilteringState(filteringState);
            return AdvancedCustomBindingCore(viewModel);
        }
        // Sorting
        public ActionResult AdvancedCustomBindingSortingAction(GridViewColumnState column, bool reset)
        {
            var viewModel = GridViewExtension.GetViewModel("gridView");
            viewModel.ApplySortingState(column, reset);
            return AdvancedCustomBindingCore(viewModel);
        }
        // Grouping
        public ActionResult AdvancedCustomBindingGroupingAction(GridViewColumnState column)
        {
            var viewModel = GridViewExtension.GetViewModel("gridView");
            viewModel.ApplyGroupingState(column);
            return AdvancedCustomBindingCore(viewModel);
        }

        PartialViewResult AdvancedCustomBindingCore(GridViewModel viewModel)
        {
            viewModel.ProcessCustomBinding(
                GridViewCustomBindingHandlers.GetDataRowCountAdvanced,
                GridViewCustomBindingHandlers.GetDataAdvanced,
                GridViewCustomBindingHandlers.GetSummaryValuesAdvanced,
                GridViewCustomBindingHandlers.GetGroupingInfoAdvanced,
                GridViewCustomBindingHandlers.GetUniqueHeaderFilterValuesAdvanced
            );
            return PartialView("_GridViewPartial", viewModel);
        }

        static GridViewModel CreateGridViewModelWithSummary()
        {
            var viewModel = new GridViewModel();
            viewModel.KeyFieldName = "OrderID";

            viewModel.Columns.Add("OrderID");
            viewModel.Columns.Add("EmployeeID");
            viewModel.Columns.Add("OrderDate");
            viewModel.Columns.Add("Freight");
            viewModel.Columns.Add("ShipName");
            viewModel.Columns.Add("ShipAddress");
            viewModel.Columns.Add("ShipCity");
            viewModel.Columns.Add("ShipCountry");

            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "OrderID", SummaryType = SummaryItemType.Sum });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "EmployeeID", SummaryType = SummaryItemType.Count });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "Freight", SummaryType = SummaryItemType.Average });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "ShipCity", SummaryType = SummaryItemType.Min });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "OrderDate", SummaryType = SummaryItemType.Min });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "Freight", SummaryType = SummaryItemType.Count });
           
            return viewModel;
        }

    }

}
